casper/src/test/scala/coop/rchain/casper/helper/DeployDataContract.scala

```diff
@@ -0,0 +1,41 @@
+package coop.rchain.casper.helper
+import cats.effect.Sync
+import coop.rchain.rholang.interpreter.{ContractCall, RhoType}
+import coop.rchain.rholang.interpreter.Runtime.SystemProcess
+import coop.rchain.models.{ListParWithRandomAndPhlos, Par}
+
+object DeployDataContract {
+  import cats.implicits._
+
+  def set[F[_]: Sync](
+      ctx: SystemProcess.Context[F]
+  )(message: (Seq[ListParWithRandomAndPhlos], Int)): F[Unit] = {
+
+    val isContractCall = new ContractCall(ctx.space, ctx.dispatcher)
+    message match {
+      case isContractCall(
+          produce,
+          Seq(RhoType.String("userId"), (pk @ RhoType.ByteArray(_)), ackCh)
+          ) =>
+        for {
+          params    <- ctx.shortLeashParams.getParams
```

Comment:
So, `set` followed by `get` on `Ref` is not atomic. You have to use `getAndSet` or `modify` for atomicity, otherwise you introduce race conditions.

Reply:
This is only used in testing but that's still a very good point. I'll fix this asap.